lib/core/utils/preload-cssom.js

```diff
@@ -1,48 +1,122 @@
+ */
+function getExternalStylesheet({
+	resolve,
+	reject,
+	url,
+	rootDocument,
+	timeout,
+	getStyleSheet,
+	shadowId,
+	priority
+}) {
+	axe.imports
+		.axios({
+			method: 'get',
+			url,
+			timeout
+		})
+		.then(({ data }) => {
+			const sheet = getStyleSheet({
+				data,
+				isExternal: true,
+				shadowId,
+				root: rootDocument,
+				priority
+			});
+			resolve(sheet);
+		})
+		.catch(reject);
+}
+
 /**
  * Returns a then(able) queue of CSSStyleSheet(s)
- * @param {Object} ownerDocument document object to be inspected for stylesheets
- * @param {number} timeout on network request for stylesheet that need to be externally fetched
- * @param {Function} convertTextToStylesheetFn a utility function to generate a style sheet from text
+ * @method loadCssom
+ * @private 
+ * @param {Object} arg an object with projects essential to load CSSOM
+ * @property {Object} arg.rootDocument document or shadowDOM root document for which to process CSSOM
+ * @property {Number} arg.rootIndex a number representing the index of the document or shadowDOM, used for priority
+ * @property {String} arg.shadowId an id if undefined denotes that given root is a shadowRoot
+ * @property {Number} timeout abort duration for network request
+ * @param {Function} getStyleSheet a utility function to generate a style sheet for a given text content
  * @return {Object} queue
- * @private
+ 
  */
-function loadCssom({ root, shadowId }, timeout, convertTextToStylesheetFn) {
+function loadCssom({
+	rootDocument,
+	rootIndex,
+	shadowId,
+	timeout,
+	getStyleSheet
+}) {
+	const q = axe.utils.queue();
+
+	let styleSheets = null;
+
 	/**
-	 * Make an axios get request to fetch a given resource and resolve
-	 * @method getExternalStylesheet
-	 * @private
-	 * @param {Object} param an object with properties to configure the external XHR
-	 * @property {Object} param.resolve resolve callback on queue
-	 * @property {Object} param.reject reject callback on queue
-	 * @property {String} param.url string representing the url of the resource to load
-	 * @property {Number} param.timeout timeout to about network call
+	 * For shadowDOM #document-fragment node, fragment.styleSheets is not reliable in most browsers (mostly Safari)
+	 * (Issue: https://github.com/dequelabs/axe-core/issues/1082)
+	 *
+	 * For stable results of CSSOM inside document fragment of shadowRoot, it is best to parse the children of the root
+	 * and extaract tags <style> and <link> to construct dynamic styleSheets
 	 */
-	function getExternalStylesheet({ resolve, reject, url }) {
-		axe.imports
-			.axios({
-				method: 'get',
-				url,
-				timeout
-			})
-			.then(({ data }) => {
-				const sheet = convertTextToStylesheetFn({
-					data,
-					isExternal: true,
-					shadowId,
-					root
+	if (
+		rootDocument.nodeName.toUpperCase() === '#DOCUMENT-FRAGMENT' &&
+		shadowId
+	) {
+		// retrieve shadowRoot style sheets as []
+		styleSheets = Array.from(rootDocument.children).reduce((out, node) => {
```

Comment:
Why are you only looking at child nodes, instead of doing getElementsByTagName for `style` and `link` elements? It might make sense to break up the function into different load methods depending on the elements. Reduces the complexity of this function.